File: lernfeld_03_oop/cpp/loesungen/aufgabe_02.cpp
// Aufgabe 2: Vererbung – Fahrzeuge — Musterlösung (C++)
//
// Basisklasse Fahrzeug (protected: marke, baujahr), Unterklassen Auto und
// Fahrrad. Der Basis-Konstruktor wird über die Initialisierungsliste
// aufgerufen (das Pendant zu Pythons super().__init__()). Statt __str__
// hat jede Klasse eine print()-Methode.
//
// Kompilieren:  g++ -std=c++17 -Wall -Wextra aufgabe_02.cpp -o aufgabe_02
// Ausführen:    ./aufgabe_02

#include <iostream>  // std::cout, std::endl
#include <string>    // std::string

class Fahrzeug {
protected:  // für Unterklassen sichtbar, nach außen aber geschützt
    std::string marke;
    int baujahr;

public:
    Fahrzeug(const std::string& marke, int baujahr)
        : marke(marke), baujahr(baujahr) {}

    void beschleunigen() const {
        std::cout << "Das Fahrzeug beschleunigt." << std::endl;
    }

    void print() const {
        std::cout << "Fahrzeug: " << marke << " (" << baujahr << ")"
                  << std::endl;
    }
};

class Auto : public Fahrzeug {
private:
    int anzahl_tueren;

public:
    Auto(const std::string& marke, int baujahr, int anzahl_tueren)
        : Fahrzeug(marke, baujahr), anzahl_tueren(anzahl_tueren) {}

    void beschleunigen() const {  // überschrieben
        std::cout << "  Das Auto beschleunigt: 0 auf 100 km/h in 9.2 s"
                  << std::endl;
    }

    void hupen() const {
        std::cout << "  Hupen: Hup Hup!" << std::endl;
    }

    void print() const {
        std::cout << "Auto: " << marke << " (" << baujahr << "), "
                  << anzahl_tueren << " Tueren" << std::endl;
    }
};

class Fahrrad : public Fahrzeug {
private:
    int gangzahl;

public:
    Fahrrad(const std::string& marke, int baujahr, int gangzahl)
        : Fahrzeug(marke, baujahr), gangzahl(gangzahl) {}

    void beschleunigen() const {  // überschrieben
        std::cout << "  Das Fahrrad beschleunigt: 0 auf 25 km/h in 8.0 s"
                  << std::endl;
    }

    void klingeln() const {
        std::cout << "  Klingeln: Kling Kling!" << std::endl;
    }

    void print() const {
        std::cout << "Fahrrad: " << marke << " (" << baujahr << "), "
                  << gangzahl << " Gaenge" << std::endl;
    }
};

int main() {
    Auto auto_objekt("VW Golf", 2018, 4);
    Fahrrad fahrrad("Giant", 2021, 21);

    auto_objekt.print();
    auto_objekt.beschleunigen();
    auto_objekt.hupen();

    std::cout << std::endl;
    fahrrad.print();
    fahrrad.beschleunigen();
    fahrrad.klingeln();

    // Auch ein reines Basis-Fahrzeug anlegen – nicht jedes Fahrzeug
    // ist ein Auto!
    std::cout << std::endl;
    Fahrzeug gabelstapler("Linde", 2015);
    gabelstapler.print();
    gabelstapler.beschleunigen();

    return 0;
}

File: lernfeld_03_oop/python/loesungen/aufgabe_02.py
"""Aufgabe 2: Vererbung – Fahrzeuge — Musterlösung (Python).

Basisklasse Fahrzeug, Unterklassen Auto und Fahrrad: gemeinsame Attribute
kommen aus der Basisklasse, Spezielles steht in der Unterklasse.
`super().__init__(...)` ruft den Basis-Konstruktor auf, `__str__` macht
print() lesbar.
"""


class Fahrzeug:
    """Basisklasse: alles, was alle Fahrzeuge gemeinsam haben."""

    def __init__(self, marke: str, baujahr: int) -> None:
        self.marke = marke
        self.baujahr = baujahr

    def beschleunigen(self) -> None:
        print("Das Fahrzeug beschleunigt.")

    def __str__(self) -> str:
        return f"Fahrzeug: {self.marke} ({self.baujahr})"


class Auto(Fahrzeug):
    """Auto: erbt marke/baujahr von Fahrzeug, hat zusätzlich Türen."""

    def __init__(self, marke: str, baujahr: int, anzahl_tueren: int) -> None:
        super().__init__(marke, baujahr)  # Basis-Konstruktor aufrufen
        self.anzahl_tueren = anzahl_tueren

    def beschleunigen(self) -> None:
        # Überschrieben: das Auto meldet sich mit eigenen Werten
        print("  Das Auto beschleunigt: 0 auf 100 km/h in 9.2 s")

    def hupen(self) -> None:
        print("  Hupen: Hup Hup!")

    def __str__(self) -> str:
        return f"Auto: {self.marke} ({self.baujahr}), " \
               f"{self.anzahl_tueren} Türen"


class Fahrrad(Fahrzeug):
    """Fahrrad: erbt marke/baujahr von Fahrzeug, hat zusätzlich Gänge."""

    def __init__(self, marke: str, baujahr: int, gangzahl: int) -> None:
        super().__init__(marke, baujahr)
        self.gangzahl = gangzahl

    def beschleunigen(self) -> None:
        print("  Das Fahrrad beschleunigt: 0 auf 25 km/h in 8.0 s")

    def klingeln(self) -> None:
        print("  Klingeln: Kling Kling!")

    def __str__(self) -> str:
        return f"Fahrrad: {self.marke} ({self.baujahr}), " \
               f"{self.gangzahl} Gänge"


def main() -> None:
    # Je ein Auto und ein Fahrrad anlegen
    auto = Auto("VW Golf", 2018, 4)
    fahrrad = Fahrrad("Giant", 2021, 21)

    # print() ruft automatisch __str__ auf
    print(auto)
    auto.beschleunigen()
    auto.hupen()

    print()
    print(fahrrad)
    fahrrad.beschleunigen()
    fahrrad.klingeln()

    # Hinweis aus der Aufgabe: Nicht jedes Fahrzeug ist ein Auto –
    # auch ein Basis-Fahrzeug-Objekt lässt sich anlegen.
    print()
    gabelstapler = Fahrzeug("Linde", 2015)
    print(gabelstapler)
    gabelstapler.beschleunigen()

    # isinstance zeigt, was "ein Fahrzeug" ist – nämlich auch Auto/Fahrrad
    print()
    print(f"auto ist ein Fahrzeug: {isinstance(auto, Fahrzeug)}")
    print(f"fahrrad ist ein Fahrzeug: {isinstance(fahrrad, Fahrzeug)}")
    print(f"gabelstapler ist ein Auto: {isinstance(gabelstapler, Auto)}")


if __name__ == "__main__":
    main()
